internal/server/boltdbstate/job.go

```diff
@@ -1772,11 +1772,25 @@ func (s *State) pipelineComplete(jobId string) error {
 	if job.State == pb.Job_ERROR {
 		run.State = pb.PipelineRun_ERROR
 	} else if job.State == pb.Job_SUCCESS {
-		// If job Id matches last job queued by pipeline.
-		// We will have to change this in the future when pipeline steps run in parallel,
-		// and the last job queued may not be the last job to complete in the pipeline
-		// TODO:XX figure out how ^
-		if job.Id == run.Jobs[len(run.Jobs)-1].Id {
+		// Look at all job ids in a run and check if any are not SUCCESS
+		runComplete := true
+		for _, j := range run.Jobs {
+			if j.Id == job.Id {
```

Comment:
Why are we skipping the current job?

Instinctively this feels really redundant to loop through every job in the run, every time we hit a job... but I can't think of a better solution off the top of my head. Maybe combine it with the existing logic so that it only hits this loop if job.Id is the last job in the queue, and _then_ make sure that every other job has completed as well?

Reply:
We skip the current job because we already know from line 1774 that the current jobs state is SUCCESS. No need to make another db call to determine this - we already have the job.

The main reason for this fix was because the `run.Jobs` slice wasn't actually in run order, the last job wasn't the final job in the pipeline, and so the run never completes and the state is stuck as `running`. Although this adds some extra overhead to look up jobs by id, I think it's an ok tradeoff until we implement the job system improvements we are planning for the next major release which will end up making this simpler anyways.

One performance improvement we could make is a specialized job state func that looks up an array of jobs by ids in a single transaction instead of per-job.

Reply:
> The main reason for this fix was because the run.Jobs slice wasn't actually in run order

🤣